hail/src/main/scala/is/hail/expr/ir/ndarrays/EmitNDArray.scala

```diff
@@ -275,56 +275,51 @@ object EmitNDArray {
-                val (pushElement, finish) = stagedArrayOfSizesPType.constructFromFunctions(cb, region, arrLength, false)
+                val concatAxisSize = cb.newLocal[Long](s"ndarray_concat_axis_size", 0L)
 
-                val newShape = (0 until x.typ.nDims).map { dimIdx =>
-                  val localDim = cb.newLocal[Long](s"ndarray_concat_output_shape_element_${dimIdx}")
-                  val ndShape = firstND.shapes
-                  cb.assign(localDim, ndShape(dimIdx))
-                  if (dimIdx == axis) {
-                    pushElement(cb, EmitCode(Code._empty, false, primitive(localDim)).toI(cb))
-                  }
+                ndsArraySValue.forEachDefined(cb) { (cb, i, nd) =>
+                  val dimLength = nd.asNDArray.shapes(axis)
+                  pushSize(cb, EmitCode(Code._empty, false, primitive(dimLength)).toI(cb))
+                  cb.assign(concatAxisSize, concatAxisSize + dimLength)
+                }
+                val stagedArrayOfSizes = finishSizes(cb)
 
-                  cb.forLoop(cb.assign(loopIdx, 1), loopIdx < arrLength, cb.assign(loopIdx, loopIdx + 1), {
-                    val shapeOfNDAtIdx = ndsArraySValue.loadElement(cb, loopIdx).map(cb) { sCode => sCode.asNDArray }.get(cb).shapeStruct(cb)
-                    val dimLength = cb.newLocal[Long]("dimLength", shapeOfNDAtIdx.loadField(cb, dimIdx).get(cb).asInt64.value)
+                // compute index of first input which has non-zero concat axis size
+                val firstNonEmpty = cb.newLocal[Int]("ndarray_concat_first_nonempty", 0)
+                cb.whileLoop(stagedArrayOfSizes.loadElement(cb, firstNonEmpty).get(cb).asInt64.value.ceq(0L), {
+                  cb.assign(firstNonEmpty, firstNonEmpty + 1)
+                })
 
-                    if (dimIdx == axis) {
-                      pushElement(cb, EmitCode(Code._empty, false, primitive(dimLength)).toI(cb))
-                      cb.assign(localDim, localDim + dimLength)
-                    }
-                    else {
-                      cb.ifx(dimLength.cne(localDim),
-                        cb._fatal(const(s"NDArrayConcat: mismatched dimensions of input NDArrays along axis ").concat(loopIdx.toS).concat(": expected ")
-                          .concat(localDim.toS).concat(", got ")
-                          .concat(dimLength.toS))
-                      )
-                    }
-                  })
-                  localDim
+                // check that all sizes along other axes are consistent
+                ndsArraySValue.forEachDefined(cb) { case (cb, _, nd: SNDArrayValue) =>
+                  val mismatchedDim = cb.newLocal[Int]("ndarray_concat_mismatched_dim", -1)
+                  val expected = cb.newLocal[Long]("ndarray_concat_expected_size")
+                  val found = cb.newLocal[Long]("ndarray_concat_found_size")
+                  for (i <- (0 until firstND.st.nDims).reverse if i != axis) {
```

Comment:
I think you can do something like `first.st.nDims - 1 to 0 by -1`, though `.reverse` is fine too

Reply:
I find it tricky to think about endpoint inclusivity using the -1 step, reverse is harder to get wrong.